std/json.c3

```diff
@@ -0,0 +1,428 @@
+function parse_element : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    cursor:eat_whitespace();
+
+    const value_node : Optional<JSON_Node> = cursor:parse_value();
+
+    cursor:eat_whitespace();
+
+    return value_node;
+}
+
+function parse_string : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    if !cursor:eat(char("\"")) {
+        return make_optional<JSON_Node>();
+    }
+
+    const start_pos : isize = cursor.position;
+
+    while !cursor:is_eof() {
+        if cursor:eat(char("\"")) {
+            const end_pos : isize = cursor.position - 1;
+
+            let node : JSON_Node = new_json_node(/* string */ 3);
+            node.value = make_optional(
+                cursor.data:slice_between(start_pos, end_pos)
+            );
+
+            return make_optional(node);
+        }
+
+        // TODO escape characters.
+
+        cursor:next();
+    }
+
+    return make_optional<JSON_Node>();
+}
+
+function is_digit_helper : (token : Optional<u8>) -> bool = {
+    return token:has_value() and ascii_is_digit(token:data());
+}
+
+function parse_number : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    const start_pos : isize = cursor.position;
+    let is_valid : bool = false;
+
+    // This is optional, so discard the return value.
+    cursor:eat(char("-"));
+
+    if !cursor:eat(char("0")) {
+        while is_digit_helper(cursor:peak()) {
+            is_valid = true;
+            cursor:next();
+        }
+    }
+    else {
+        is_valid = true;
+    }
+
+    // TODO parse fraction.
+    // TODO parse exponent.
+
+    const end_pos : isize = cursor.position;
+
+    if !is_valid {
+        return make_optional<JSON_Node>();
+    }
+
+    let node : JSON_Node = new_json_node(/* string */ 4);
+    node.value = make_optional(cursor.data:slice_between(start_pos, end_pos));
+
+    return make_optional(node);
+}
+
+// TODO this and write below should be the only public functions.
+function json_parse : (source : StringView) -> Optional<JSON_Node> = {
+    // FIXME memory safety, memory leaks.
+    let cursor : Cursor = {
+        source, 0, initialize_allocator()
+    };
+
+    let result : Optional<JSON_Node> = cursor:parse_element();
+
+    if !result:has_value() {
+        return result;
+    }
+
+    // Only succeed if we have consumed the entire input.
+    // e.g. '[1, 2], 2' would parse up to the end of the array, but the
+    // remaining input makes this invalid json.
+    cursor:eat_whitespace();
+    if cursor:is_eof() {
+        return result;
+    }
+
+    return make_optional<JSON_Node>();
+}
+
+function [IO] write : (io : IO&, node : JSON_Node) -> TypeIf<StringView, IsStreamableTextIO<IO>> = {
+    // null
+    if node.type == 0 {
+        return io:write(sv("null"));
+    }
+
+    // true
+    if node.type == 1 {
+        return io:write(sv("true"));
+    }
+
+    // false
+    if node.type == 2 {
+        return io:write(sv("false"));
+    }
+
+    // string
+    if node.type == 3 {
+        // Allowed to fail.
```

Comment:
Why are these allowed to fail?

Reply:
Cause we can't return an error! Crashing the whole program sounds pretty excessive tbh